lib/ImageMatrix.cpp

```diff
@@ -121,4 +121,71 @@ namespace CIL {
         }
         *this = new_img;
     }
+
+    void ImageMatrix::forEach(
+        std::function<void(const typename ImageMatrix::ValueType&)> fn) const
+    {
+        for (auto i = 0U; i < m_height; ++i)
+        {
+            for (auto j = 0U; j < m_width; ++j)
+            {
+                for (auto k = 0U; k < m_num_components; ++k)
+                {
+                    fn((*this)(i, j, k));
+                }
+            }
+        }
+    }
+
+    void ImageMatrix::forEach(
+        std::function<void(const typename ImageMatrix::ValueType&, uint32_t row,
+                           uint32_t col, uint32_t comp)>
+            fn) const
+    {
+        for (auto i = 0U; i < m_height; ++i)
+        {
+            for (auto j = 0U; j < m_width; ++j)
+            {
+                for (auto k = 0U; k < m_num_components; ++k)
+                {
+                    fn((*this)(i, j, k), i, j, k);
+                }
+            }
+        }
+    }
+
+    ImageMatrix computeAbsDifference(const ImageMatrix& data1,
+                                     const ImageMatrix& data2)
+    {
+        assert(data1.width() == data2.width() &&
+               data1.height() == data2.height() &&
+               data1.numComponents() == data2.numComponents() &&
+               data1.sampleDepth() == data2.sampleDepth());
+        ImageMatrix data(data1.width(), data2.height(), data1.numComponents(),
+                         data1.sampleDepth());
+        data1.forEach([&data, &data2](const typename ImageMatrix::ValueType val,
+                                      uint32_t row, uint32_t col,
+                                      uint32_t comp) {
+            data(row, col, comp) = std::abs(val - data2(row, col, comp));
```

Comment:
I think this can be error prone, the subtraction inside this `abs` has unsigned datatypes which will never give negative results. We need to cast it to signed ints I think. Am I wrong?

Reply:
Good point. Nope, you are not missing anything. I will modify the code to use casts here.